Padmiss support

electromuis opened the pull request:
With thanks to the finnish ITG community

Add support for folder based MemoryCard usage
Add support fot the https://padmiss.com/ network by saving scores to individual files.

(For communication with padmiss, use https://github.com/electromuis/padmiss-daemon)

dguzek commented:
Thanks for this pull request!

I'm completely unfamiliar with padmiss, and the website didn't seem to offer any front-facing information about how it is used.  Additionally, padmiss-daemon doens't have a readme or anything similar from which one new to the project might glean what it does or how to use it.

Can you perhaps write a little bit more about this pull request?  What problem does it solve?  How does it accomplish that?  If merged, how would it be used by StepMania players?

electromuis commented:
Its an online scoretracker which has been used in the past few ITG Eurocop events.
The flow works like:
- Cab admin a padmiss token (one to one case by now, still in development)
- Cab admin updates stepmania, and sets these lines:
MemoryCardOsMountPointP1=/media/dummy1 (or whatever)
MemoryCardOsMountPointP2=/media/dummy2 (or whatever)
MemoryCardDriver=Directory
MemoryCardPadmissEnabled=1
- Cab admin installs the daemon and configures it with the scanners
 - Daemon hooks into nfc reader modules (Search "rfid reader usb" on aliexpress/ebay) which normally act like a keyboard
- Player get a rfid card, depending on what scanner they use
- They register it over at https://padmiss.com/register
 - Player scans their card when you would normally instert your usb
 - The daemon reads the code, finds the player via de padmiss api. Then it makes a folder where stepmania expects the MemoryCard with all user relevant user data
 - Players plays a song
 - The StatsManaget saves an entry with the song/score/player data to a file in Save/Padmiss
 - The daemon notices a change in that folder,
 - The daemon reads the new file, and submits it to the api
 - The players opens the website or app and notices their new score

The daemon still needs an easily installable version, with that some documentation on configuration will come as well

electromuis commented:
Windows build including these changes: https://drive.google.com/open?id=1_VFzXZg-fO4fxmYPV-k0ZCIalmlXM2LX

Horsey- commented:
So when you say "app" can the app have NFC-enabled phones broadcast your RFID "card" to the reader? YEARS AGO there was a few apple engineers from the U of A that made a hook for Apple Wallet to display a login code for the same purpose (this never took off for whatever reason, mostly because at the time it was much more difficult to get your own private machine).

electromuis commented:
With app i mean: https://github.com/electromuis/padmiss-app
Which is just another interface for the website.

@horsey i suppose you mean using your phone as your card? Depending on the scanners that could work.
Padmiss hasn't really settled on a standard (which might not even happen). For now we just use the cheapest option

Alhetus commented:
Hi,

Some clarifications:
- the card readers have nothing to do with nfc (for now), they are the most basic usb rfid readers you can find on eBay, because 3$ is a really nice price point
- at this point Padmiss is pretty much just an API with a very basic and hacky react front, but the target is to improve the front greatly in the future to host tournaments and track global scores with player levels, achievements etc.
- anyone can basically create any player data fetching method for Padmiss if you just know the player rfidUid (a.k.a "secret", that can be any string really) that you post to the API.
- a "cab" is currently in charge of posting the scores to the API with a designated API key, I have future plans to actually have the player log in more securely (credentials or rfid card + pin code) and post the scores with a player token instead

coderjo commented:
I have a question. first, do I have the sequence of operations correct here?

Upon scanning a card and successfully retrieving the info from the padmiss API, it creates the profile directory and populates it with a very basic stats.xml and editable.ini.

The MCDT_Folder class notices it in the memory card worker thread polling and tells MCM there is a card there, which the game loads.

Then the game, after each stage, writes stats in a special subdirectory of the profile, which the daemon notices and sends to padmiss.

After the game ends, how does the profile directory get removed? It appears that it must be at some point, because the MCDT_Folder code only appears to check if the profile folder exists, not that the profile data has changed.

I'm guessing this was only used on linux and/or macos so far, and not windows, correct? I think this might need minor tweaks for windows (the stat function is named _stat on windows, as is the struct).

Oh, and you should probably change the copyright on MemoryCardDriverThreaded_Folder.*, since it's your code. You just started with a class structure to implement.

electromuis commented:
@coderjo 
The daemon clears the folder when an unknown card is scanned (for EC we used empty cards and marked them: Check out)
Also, when someone is checked and then someone else scans the card it removes the existing folder, and loads the new profile.

The goal is to also check out when the same card is scanned twice, but this is not implemented as of speaking.

Windows is supported! (At least it worked like this on my w10 machine) But for now, only 1 reader can be used. Using multiple is on the schedule though. 

As for copyright, i ll change that with my next update. Also want to notify stepmania when the last modification date changes

electromuis commented:
Updated the copyright description,
Also added a lua function that allows song filtering from the theme. In my case i use it to hide certain packs in a Blacklist.txt. The packs become shown when a keycode is used

shakesoda commented on src/NotesWriterSM.cpp:
```diff
@@ -376,4 +381,4 @@ bool NotesWriterSM::WriteEditFileToMachine( const Song *pSong, Steps *pSteps, RS
  * OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
  * PERFORMANCE OF THIS SOFTWARE.
- */
+ */
```
please keep newlines at end, not having them causes warnings.

shakesoda commented on src/NotesWriterSM.h:
```diff
@@ -3,45 +3,47 @@
+     * @param sPath the path to write the file.
+     * @param out the Song to be written out.
+     * @return its success or failure. */
+    bool Write( RageFileBasic &file, Song &out, const vector<Steps*>& vpStepsToSave );
```
fix style, we indent with tabs

shakesoda commented on src/NotesWriterSM.h:
```diff
@@ -61,4 +63,4 @@ namespace NotesWriterSM
  * OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
  * PERFORMANCE OF THIS SOFTWARE.
- */
+ */
```
add newline

shakesoda commented on src/StatsManager.cpp:
```diff
@@ -227,60 +234,237 @@ void StatsManager::CommitStatsToProfiles( const StageStats *pSS )
+	ADD_BOOLEAN_OPTION( transforms, TRANSFORM_NOSTRETCH, opts.m_bTransforms );
+
+#define ADD_FLOAT_OPTION( parent, name, opts ) \
+    do { \
```
space -> tab

shakesoda commented on src/StatsManager.cpp:
```diff
@@ -458,4 +642,4 @@ LUA_REGISTER_CLASS( StatsManager )
  * OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
  * PERFORMANCE OF THIS SOFTWARE.
- */
+ */
```
add newline

shakesoda commented on src/StatsManager.h:
```diff
@@ -7,36 +7,40 @@
 public:
-	StatsManager();
-	~StatsManager();
+    StatsManager();
```
change to tab

shakesoda commented on src/StatsManager.h:
```diff
@@ -68,4 +72,4 @@ extern StatsManager*	STATSMAN;	// global and accessible from anywhere in our pro
  * OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
  * PERFORMANCE OF THIS SOFTWARE.
- */
+ */
```
add newline

shakesoda commented on src/UnlockManager.cpp:
```diff
@@ -412,6 +412,42 @@ UnlockEntryStatus UnlockEntry::GetUnlockEntryStatus() const
+
+	Lua *L = LUA->Get();
+	songEnabled.PushSelf(L);
+    pSong->PushSelf(L);
```
change to tabs

shakesoda commented on src/arch/MemoryCard/MemoryCardDriver.cpp:
```diff
@@ -7,6 +7,23 @@
+};
+
+static const char *MemoryCardDriverTypeNames[] = {
+		"USB",
```
double indented

shakesoda commented on src/arch/MemoryCard/MemoryCardDriverThreaded_Folder.cpp:
```diff
@@ -0,0 +1,131 @@
+
+bool MemoryCardDriverThreaded_Folder::TestWrite( UsbStorageDevice* pDevice )
+{
+	//TODO
```
I won't block the merge on this but it'd be nice if this todo gets sorted out in the near future :)

shakesoda commented on src/arch/MemoryCard/MemoryCardDriverThreaded_Folder.cpp:
```diff
@@ -0,0 +1,131 @@
+{
+    int ret = 0;
+
+    FOREACH_PlayerNumber( p )
```
change to tabs

shakesoda commented on src/arch/MemoryCard/MemoryCardDriverThreaded_Folder.cpp:
```diff
@@ -0,0 +1,131 @@
+	return true;
+}
+
+int MemoryCardDriverThreaded_Folder::GetActivePlayers()
```
given that this function returns a bitmask, it should probably be renamed to GetActivePlayerMask()

shakesoda commented on src/arch/MemoryCard/MemoryCardDriverThreaded_Folder.cpp:
```diff
@@ -0,0 +1,131 @@
+	LOG->Trace( "GetUSBStorageDevices" );
+
+	vDevicesOut.clear();
+    m_LastDevices = GetActivePlayers();
```
change to tabs

coderjo commented:
I'd suggest putting the SongEnabled commit in a separate PR, as it is a separate feature from support for padmiss.

electromuis commented:
Sorry for the late response, appllied codestyle fixes. Also removed the song filter function to put in a seperate request

shakesoda commented:
thanks, looks good to me 👍

electromuis commented:
It shouldn't have, first i rolled back 1 commit. Then did the fixes
